ministryofjustice/analytics-platform-ops
# [WIP] Improve environment creation

## Background

At the moment this is a fiddly process which also requires manual
intervention. Because create a new environment doesn't happen often
this makes the process even more time consuming and error-prone.


## What and why

We want to automate as much as we can and reduce the need for documentation
to the minumum.

## Scripts

### create_environment.sh (main script)

The `create_environment.sh` receives the new environment name and:
1) Create the terraform files and applies they, creating the environment
VPCs, S3 buckets, etc... by running `create_tf_resources.sh`
2) Create the kubernetes cluster by generatign the KOPS resources and
applying them. This will create the K8s masters/nodes, etc...
3) TODO: Helm charts installation is next

### create_tf_resources.sh

This scripts uses the `dev` environment as template, copying the
`infra/terraform/environments/dev` directory and changing the environment
name.

The terraform resources will be places in `infra/terraform/environments/`
as usual.

### create_k8s_cluster.sh

This scripts will render the `infra/kops/example-cluster` KOPS templates
by using values extracted by the terraform environment previously created
(by script above for example)

It mainly gets values by using the `terraform output ...` functionality.

The templates are rendered by using a command line tool to render Go templates
called [gucci](https://github.com/noqcks/gucci). This tools is installed by
the script but [Go needs to be installed](https://golang.org/doc/install#install) on the system.

The script uses [jq](https://stedolan.github.io/jq/) to extract some of
the required values so it needs to be installed on your system.

An SSH key for the cluster will be generated by this script in `~/.ssh`.

The KOPS resources will be places in `infra/kops/clusters/` as usual.





## How to review

0. ([Install Go](https://golang.org/doc/install) if needed)
1. Run `infra/create_environment.sh $YOUR_ENV_NAME`
2. The script should create your environment
    1. Create the terraform resources under `infra/terraform/environments/$YOUR_ENV_NAME`
    2. Create the basic infrastructure by running `terraform apply`
    3. Create the KOPS resources under `infra/kops/clusters/$YOUR_ENV_NAME`
    4. Create the kubernetes cluster by running `kops update cluster ...`
3. Check the k8s cluster is created, e.g. log into AWS Console and see if the EC2 instances for master/nodes were created and run `kubectl clusterinfo`.


## Ticket

https://trello.com/c/huzsYGGB/251-improve-terraform-kops-bash-resources-to-more-easily-deploy-simplified-environments-for-individual-dev-and-test-envs
